lib/src/tag_detection.dart

```diff
@@ -0,0 +1,372 @@
+
+  @override
+  Set<Uri> directSuccessors(Uri uri) {
+    return _cache.putIfAbsent(uri, () {
+      if (uri.toString().startsWith('dart:')) {
+        return <Uri>{};
+      }
+      final unitResult = _analysisSession.getParsedUnit(
+        _analysisSession.uriConverter.uriToPath(uri),
+      );
+      if (unitResult.isPart || unitResult.unit == null) {
+        // Part files cannot contain import/export directives.
+        return null;
+      }
+      final dependencies = <Uri>{};
+      for (final node in unitResult.unit.sortedDirectivesAndDeclarations) {
+        if (node is! ImportDirective && node is! ExportDirective) {
+          continue;
+        }
+        // The syntax for an import:
+        //
+        //    directive ::=
+        //        'import' stringLiteral (configuration)*
+        //      | 'export' stringLiteral (configuration)*
+        //
+        //    configuration ::= 'if' '(' test ')' uri
+        //
+        //    test ::= dottedName ('==' stringLiteral)?
+        //
+        // We have dependency upon `directive.uri` resolved relative to this
+        // library `uri`. Unless there is a `configuration` for which the `test`
+        // evaluates to true.
+        final directive = node as NamespaceDirective;
+        var dependency = uri.resolve(directive.uri.stringValue);
+
+        for (final configuration in directive.configurations) {
+          final dottedName =
+              configuration.name.components.map((i) => i.name).join('.');
+
+          final testValue = configuration.value?.stringValue ?? 'true';
+          final actualValue = _declaredVariables[dottedName] ?? 'false';
+
+          if (actualValue == testValue) {
+            dependency = uri.resolve(configuration.uri.stringValue);
+            break; // always pick the first satisfied configuration
+          }
+        }
+
+        //TODO: Determine if there is any need to normalize the `dependency` Uri.
+        dependencies.add(dependency);
+      }
+      return dependencies;
+    });
+  }
+}
+
+class LibraryPackage {
+  final AnalysisSession _analysisSession;
+  LibraryPackage(this._analysisSession);
+  String _packageDir(Uri uri) {
+    final packageOnlyUri =
+        uri.replace(path: uri.path.substring(0, uri.path.indexOf('/') + 1));
+    return path
+        .dirname(_analysisSession.uriConverter.uriToPath(packageOnlyUri));
+  }
+
+  final Map<String, Pubspec> _pubspecCache = <String, Pubspec>{};
+
+  Pubspec pubspecOfLibrary(Uri uri) {
+    return pubspecOfPackage(_packageDir(uri));
+  }
+
+  Pubspec pubspecOfPackage(String packageDir) {
+    return _pubspecCache.putIfAbsent(packageDir, () {
+      return Pubspec.parseFromDir(packageDir);
+    });
+  }
+}
+
+class Runtime {
+  final String name;
+  final Set<String> enabledLibs;
+
+  Runtime(this.name, this.enabledLibs);
+
+  Map<String, String> get declaredVariables =>
+      {for (final lib in enabledLibs) 'dart.library.$lib': 'true'};
```

Comment:
Need to make the `dart.library.js` to `false` when not present? or is that not how it works? I realize that our logic is largely ignorant of this.